tests/Unit/Shipping/PostcodeRangeTest.php

```diff
@@ -0,0 +1,33 @@
+<?php
+declare( strict_types=1 );
+
+namespace Automattic\WooCommerce\GoogleListingsAndAds\Tests\Unit\Shipping;
+
+use Automattic\WooCommerce\GoogleListingsAndAds\Shipping\PostcodeRange;
+use Automattic\WooCommerce\GoogleListingsAndAds\Tests\Framework\UnitTest;
+
+/**
+ * Class PostcodeRangeTest
+ *
+ * @package Automattic\WooCommerce\GoogleListingsAndAds\Tests\Unit\Shipping
+ */
+class PostcodeRangeTest extends UnitTest {
```

Comment:
It looks like each postcode is mapped through this range class but we aren't actually testing ranges. Should we be testing single postcode ranges here as well? Or what about a postcode containing a wildcard?

Reply:
The comment was more intended for the `PostcodesRateGroupAdapterTest` where it only tests mapping with a postcode range. Which I suppose is fine since `PostcodeRangeTest`, tests a single range. I was also a bit confused about why it tests a postcode range for `DE` (I was under the impression that subdivisions was only supported for some countries). I guess it's just the API that has this restriction, and the mapping should work regardless.

I was originally under the impression that the class `Google\Service\ShoppingContent\PostalCodeRange` would also handle syntax checking, but it seems it just allows any value for begin and end, and it's only when it's passed to the API that it might complain about invalid syntax. I was looking for the limitations specified [here](https://developers.google.com/shopping-content/reference/rest/v2.1/regions#PostalCodeRange), while both begin and end can be a string it mentions specific formatting.

I mainly say this is missing, because WooCommerce does not prevent you from saving a postcode rate like `1234?`, Internally it will never match that location when calculating shipping. However I tested that with the mapping and it doesn't complain since it accepts all strings, but when we try to sync that to Google it responds with the following error:
```
{
	"code": "gla_setting_sync_error",
	"message": "{\n  \"error\": {\n    \"code\": 400,\n    \"message\": \"[postalCodeGroups[0].postalCodeRanges[3]] None of the pattern matched by postal code: 1234?.\",\n    \"errors\": [\n      {\n        \"message\": \"[postalCodeGroups[0].postalCodeRanges[3]] None of the pattern matched by postal code: 1234?.\",\n        \"domain\": \"global\",\n        \"reason\": \"invalid\"\n      }\n    ]\n  }\n}\n",
	"data": {
		"status": 500,
		"message": "",
		"data": {
			"error": {
				"code": 400,
				"message": "[postalCodeGroups[0].postalCodeRanges[3]] None of the pattern matched by postal code: 1234?.",
				"errors": [
					{
						"message": "[postalCodeGroups[0].postalCodeRanges[3]] None of the pattern matched by postal code: 1234?.",
						"domain": "global",
						"reason": "invalid"
					}
				]
			}
		}
	}
}
```